File: 653-two-sum-iv-input-is-a-bst/653.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def traverse(self, node):
        if node is None:
            return False

        self.traverse(node.left)
        self.traverse(node.right)

        self.arr.append(node.val)

    def find(self, root, n, target):
        if root is None:
            return False

        if root.val == target and root.val != n:
            return True

        if root.val > target:
            return self.find(root.left, n, target)
        else:
            return self.find(root.right, n, target)

    def findTarget(self, root: Optional[TreeNode], k: int) -> bool:
        self.arr = []

        self.traverse(root)

        for n in self.arr:
            if self.find(root, n, k-n):
                return True

        return False

File: 653-two-sum-iv-input-is-a-bst/653.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
private:
    bool dfs(TreeNode* root, TreeNode* node, int k) {
        if (!root || !node)
            return false;
        int diff = k - node->val;
        if (diff != node->val) {
            TreeNode* curr = root;
            while (curr) {
                if (diff == curr->val)
                    return true;
                if (diff > curr->val)
                    curr = curr->right;
                else
                    curr = curr->left;
            }
        }
        bool left = dfs(root, node->left, k);
        bool right = dfs(root, node->right, k);
        return left || right;
    }
public:
    bool findTarget(TreeNode* root, int k) {
        return dfs(root, root, k);
    }
};
